source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,467 @@
+                Ok(value)
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+    }
+
+    /// Toggle the card between stand-by and transfer state
+    pub async fn select(&mut self, rca: u32) -> Result<CardStatus, Error> {
+        match self
+            .cmd(Command {
+                index: 7,
+                argument: rca,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD7 response: {value:#x}");
+                Ok(CardStatus(value))
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+    }
+
+    /// Use 4 data lanes
+    pub async fn set_wide_bus(&mut self, rca: u32) -> Result<CardStatus, Error> {
+        // Go to *APP* mode before sending application command
+        self.cmd(Command {
+            index: 55,
+            argument: rca,
+            options: HardwareOptions::None,
+            kind: CommandKind::Control,
+            rsp_type: ResponseType::Short,
+            rsp_crc: true,
+            buffer: None,
+        })
+        .await?;
+
+        match self
+            .cmd(Command {
+                index: 6,
+                argument: 0b10,
+                options: HardwareOptions::SetBusWidth(BusWidth::Quad),
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("ACMD6 response: {value:#x}");
+                Ok(CardStatus(value))
+            }
+            Response::Long(_) => Err(Error::Response),
+        }
+    }
+
+    /// Read the desired number of data blocks into the provided buffer, starting at the given sector.
+    pub async fn read(
+        &mut self,
+        sector: u32,
+        blocks: usize,
+        buf: FixedVec<u8>,
+    ) -> Result<FixedVec<u8>, Error> {
+        match self
+            .cmd(Command {
+                index: 18,
+                argument: sector,
+                options: HardwareOptions::None,
+                kind: CommandKind::Read(512 * blocks),
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: Some(buf),
+            })
+            .await?
+        {
+            Response::Short {
+                value,
+                data: Some(res),
+            } => {
+                tracing::trace!("CMD18 response: {value:#x}");
+                Ok(res)
+            }
+            _ => Err(Error::Response),
+        }
+    }
+}
+
+/// A client for SDIO cards using the [`SdmmcService`].
+pub struct SdioClient {
```

Comment:
let's maybe add 
```suggestion
#[allow(dead_code)]
pub struct SdioClient {
```
on this for now so that clippy stops whining about it?